HIP/hip-514-missing-hts-precompile-functions.md

```diff
@@ -0,0 +1,243 @@
+
+As a smart contract developer, I would like to retrieve token key properties through a precompile contract.
+
+As a smart contract developer, I would like to delete a token through a precompile contract.
+
+As a smart contract developer, I would like to wipe an amount of fungible tokens from an account balance through a precompile contract.
+
+As a smart contract developer, I would like to wipe an amount of non-fungible tokens from an account balance through a precompile contract.
+
+## Specification
+
+### EVM Precompile extension
+
+The [IHederaTokenService.sol Solidity file](../assets/hip-206/solidity/IHederaTokenService.sol) is further extended with new types and function signatures that the precompile will respond to. 
+
+### New Solidity Types
+
+The following structs have been added to simplify the interface between Solidity and the HTS precompile: 
+
+| Name                      | Definition                                                            |
+|---------------------------|-----------------------------------------------------------------------|
+| `CustomFee`               | `(FixedFee, FractionalFee, RoyaltyFee)`                               |
+| `TokenInfo`               | `(HederaToken, uint64, bool, bool, bool, CustomFee[], string)`        |
+| `FungibleTokenInfo`       | `(TokenInfo, uint32)`                                                 |
+| `NonFungibleTokenInfo`    | `(TokenInfo, int64, address, uint32, bytes, address)`                 |
+
+The additional structs build upon existing structs and in most add network info (e.g. ledgerId) or differentiating token information (e.g. fungible vs non-fungible).
+
+#### CustomFee
+`CustomFee` struct defines the fees collected and the collection account when transferring a Hedera Token:
+
+| Field                             | Meaning                               |
+|-----------------------------------|---------------------------------------|
+| `FixedFee`                        | The fixed fee for the token           |
+| `FractionalFee`                   | The fractional fee for the token      |
+| `RoyaltyFee`                      | The royalty fee for the token         |
+
+* Note that a valid `CustomFee` will have **only one** valid fee i.e. `FixedFee` or `FractionalFee` or `RoyaltyFee`.
+
+#### TokenInfo
+`TokenInfo` struct defines the basic properties of a Fungible or Non-Fungible Hedera Token:
+
+| Field                     | Meaning                                                                                           |
+|---------------------------|---------------------------------------------------------------------------------------------------|
+| `HederaToken token`       | The hedera token.                                                                                 |
+| `uint64 totalSupply`      | The number of tokens (fungible) or serials (non-fungible) of the token                            |
```

Comment:
In `MerkleToken` we keep `totalSupply` as `long` which covers the range of `int64` in solidity. Probably we can use `int64` here, since `uint64` might cause overflow issues.

Reply:
Actually, totalSupply is used only for returning read-only info and it is safe to encode value to a bigger namespace (long/int64 to uint64). The opposite action is more dangerous (decoding uint64 to a smaller range of values int64/long). So, to reduce effort we can leave totalSupply as uint64 in all places.

Reply:
Updated and reverted :)